walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,72 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
+  /**
+   * Finds a collection of time range that the meeting can be hold
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Sort the events list based the starting time of each event
+    Event[] eventsArray = events.toArray(new Event[0]);
+    Arrays.sort(eventsArray, ORDER_BY_TIME_RANGE);
+
+    // Find the collection of time range to hold the meeting
+    Collection<TimeRange> queries = new ArrayList<>();
+    int preStartTime = TimeRange.START_OF_DAY;
+    for (Event event : eventsArray) {
+      TimeRange eventTimeRange = event.getWhen();
+      if (!areAllAttendeesFree(event, request)) {
+        if (canRequestBeHold(preStartTime, eventTimeRange.start(), request)) {
+          queries.add(TimeRange.fromStartEnd(preStartTime, eventTimeRange.start(), false));
+        }
+        // Update the preStartTime to point next possible time to hold the meeting
+        int eventEndTime = eventTimeRange.end();
+        preStartTime = eventEndTime > preStartTime ? eventEndTime : preStartTime;
+      }
+    }
+    // Take care the END_OF_DAY to see if the meeting can be hold
+    if (canRequestBeHold(preStartTime, TimeRange.END_OF_DAY, request)) {
+      queries.add(TimeRange.fromStartEnd(preStartTime, TimeRange.END_OF_DAY, true));
+    }
+    return queries;
+  }
+
+  /**
+   * Returns true if the event contains people that are required to attend the metting
+   * false otherwise.
+   */
+  private static boolean areAllAttendeesFree(Event event, MeetingRequest request) {
+    Collection<String> eventAttendees = event.getAttendees();
```

Comment:
consider replace this function with [Collections.disjoint](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#disjoint-java.util.Collection-java.util.Collection-)?

Reply:
Done!